### PR TITLE
[python-package] add type hints on some functions in basic.py

### PR DESCRIPTION
**Pull request by makquel**
Contributes to #3756.

Update static type for the following functions that belong to the basic package:

- cfloat32_array_to_numpy
- cfloat64_array_to_numpy
- cint32_array_to_numpy
- cint64_array_to_nump
- c_str
- c_array

### CODE REVIEW

**Comment by jameslamb** [comment]
Thanks very much @makquel ! I just updated the title and description of this PR.

In the future, when you're working on something that has already been discussed in another place, please link to that discussion so maintainers understand *why* you're proposing these changes. For example, I've added "Contributes to #3756".

For pull request titles in this project, please make them the type of entry you'd expect to find in release notes, not a branch name. In this project, pull request titles become release note items. See https://github.com/microsoft/LightGBM/releases/tag/v3.3.1 for an example.

File: .pre-commit-config.yaml
Comment (jameslamb) [inline .pre-commit-config.yaml:1]: please remove this file

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks @makquel !

This pull request is not yet ready to review. It looks like maybe you've run `black` over `python-package/lightgbm/basic.py`, which has resulted in significant unrelated changes.

image

This project does not use `black`. Please revert all of those changes, so that the only thing left in this pull request are the changes related to type hinting.

**Comment by makquel** [comment]
@jameslamb I reverted black format changes, removed the pre-commit file, and finished the proposed scope.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [inline python-package/lightgbm/basic.py:277]:
```suggestion
def c_array(ctype: type, values: List[ctypes.c_char_p]) -> ctypes.Array:
```

This function is only called in one place in `lightgbm`.

https://github.com/microsoft/LightGBM/blob/273c9b043e11b24a95e40f99ef81c4888d2089c9/python-package/lightgbm/basic.py#L2178

Found with the following

```shell
git grep -E 'c_array\('
```

In that place, it's called with type `ctypes.c_char_p` (not an *instance* of that type) and provided a list of `ctypes.c_char_p` instances.

Here's some code I used to figure out the return type of this function.

```python
import ctypes

values = [
    ctypes.c_char_p("a".encode('utf-8')),
    ctypes.c_char_p("b".encode('utf-8')),
    ctypes.c_char_p("c".encode('utf-8'))
]

x = (ctypes.c_char_p * len(values))(*values)
print(type(x).__mro___)
isinstance(x, ctypes.Array)
```

Please make these changes, and for the purpose of this PR please revert the changes you've made to the function's docstring. Those changes contain a few inaccurate statements, and are not necessary for the scope of this PR.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much. Please see my next set of suggestions.

File: python-package/lightgbm/basic.py
Comment (makquel) [reply python-package/lightgbm/basic.py]: You're right, I made a similar test, however wasn't sure if `ctype.Array` was the right one of typing. 
I believe this was the last revision issue. Let me know if tou see something unusual.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [inline python-package/lightgbm/basic.py:2588]: can you please remove this and the changes to `update()` and `__boost()`? Those changes are from #5359, which has already been merged to `master`, so I'm not sure why they're showing up in the diff.

The easiest way to remove them is probably to merge `master` into your branch.

```shell
git checkout feature/static-typing
git remote add upstream https://github.com/microsoft/LightGBM.git
git pull upstream master
git push origin feature/static-typing
```

**Review by jameslamb — CHANGES_REQUESTED** [review]

File: python-package/lightgbm/basic.py
Comment (makquel) [reply python-package/lightgbm/basic.py:2588]: It's done!

Comment (jameslamb) [reply python-package/lightgbm/basic.py:2588]: I don't believe so. These lines are still showing up in the diff.

for example:

image

You may need to do something like the following.

```shell
git rebase master
git push origin feature/static-typing --force
```

Comment (makquel) [reply python-package/lightgbm/basic.py:2588]: Something weird happened the first time, but I think it's already fixed
Screen Shot 2022-07-16 at 17 38 42
.

Comment (jameslamb) [reply python-package/lightgbm/basic.py:2588]: If you navigate to https://github.com/microsoft/LightGBM/pull/5362/files in your browser, you will see that chaanges too `__boost()` and `update()` are still present in the diff.

You can see that in the statistics as well.

image

I'd expect this to be `+6, -6` since you've made 6 one-line changes.

If rebasing isn't working or you aren't sure how to do it, I'd be happy to help! If you go to https://github.com/makquel/LightGBM/settings/access and grant me write access on your fork, I can do this git work and and push changes to your branch.

Comment (makquel) [reply python-package/lightgbm/basic.py]: I made it manually, so I believe it's finally fixed!

Comment (jameslamb) [reply python-package/lightgbm/basic.py]: looks great now, thanks very much for working through that!

**Review by jameslamb — APPROVED** [review]
Looks good, thanks very much for doing this!

I'll merge this if the CI jobs run successfully. If you're open to it, we'd love more help from you on #3756.

**Comment by jameslamb** [comment]
hmmm I'm not sure what's happening, but Azure DevOps is not making progress on running the Linux jobs for this PR.

It's been stuck in this state for several hours, despite there not being any other jobs running:

image

I don't see any warnings in the logs ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13089&view=results)).

@shiyu1994 can you please check this? Maybe you received an email about a problem with our account, like hitting service limits?

**Comment by makquel** [comment]
> Looks good, thanks very much for doing this!
> 
> I'll merge this if the CI jobs run successfully. If you're open to it, we'd love more help from you on #3756.

@jameslamb I'll probably work on the [3867](https://github.com/microsoft/LightGBM/issues/3867) issue.

**Comment by jameslamb** [comment]
@makquel could you please try pushing an empty commit to this branch?

```sh
git commit --allow-empty -m "empty commit"
git push origin feature/static-typing
```

Maybe this Azure CI build got into a weird state that can't be resolved by me clicking "retry". Sorry for the extra work.

**Comment by shiyu1994** [comment]
All subsequent PR after this one did not trigger the running of CI tasks on Linux machine. For example,
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13090&view=results

**Comment by jameslamb** [comment]
ah yeah you're right, I can see that builds from other PRs are also stuck in queued.

Thanks for looking into it! I don't have access to see or manage the agents in the `sh-ubuntu` pool.

**Comment by shiyu1994** [comment]
It seems that the client secret of the application for lightgbm-ci in Azure Portal has been expired in Jul 16th, which coincides with the time when the agent goes off.
client
Unfortunately, I have no access to this application. I'm contacting the administrative team for help.

**Comment by jameslamb** [comment]
oh yikes! Ok thanks for looking into that @shiyu1994 . I hope they can give you admin access and register your email address as the owner, so we'll be notified of such things in the future before they expire.

**Comment by jameslamb** [comment]
@shiyu1994 could you please check your email and let me know if you got any response about how to fix Azure DevOps?

I'd love to merge some of the open, approved PRs over the weekend if possible, but don't want to do that if we can't test them in CI.

**Comment by shiyu1994** [comment]
@jameslamb Sorry but there's no response so far. I think we may reopen a virtual machine in our Azure Subscription to replace the existing Linux machine for CI test in Azure DevOps.

**Comment by shiyu1994** [comment]
I'll fix this within this weekend. It is blocking the merging of all the PRs.

**Comment by jameslamb** [comment]
ok thanks very much for checking!

File: python-package/lightgbm/basic.py
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:277]: I believe this is universal function, not just for strings.
For example,
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L371-L377
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L1427
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L1552
```suggestion
def c_array(ctype: type, values: List[Any]) -> ctypes.Array:
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]

File: python-package/lightgbm/basic.py
Comment (jameslamb) [reply python-package/lightgbm/basic.py:277]: hmmm ok. I recommended `ctypes.c_char_p` because the only place that it is used in `lightgbm` is with a list of `ctypes.c_char_p`, like I mentioned in https://github.com/microsoft/LightGBM/pull/5362#discussion_r918444042.

I think that most accurately it is that `"`ctype` needs to be one of the pointer types exported by `ctypes`, and `values` should be a list of corresponding instances that those pointers can refer to`.

But I understand that since there is no base class for those types in the `ctypes` namespace (just the internal `_ctypes.PyCSimpleType`), `Any` might be the best we can do.

```python
from ctypes import c_char_p
type(c_char_p)
# <class '_ctypes.PyCSimpleType'>

type(c_char_p).__mro__
# (<class '_ctypes.PyCSimpleType'>, <class 'type'>, <class 'object'>)
```

I'm ok with this change. @makquel will you please accept this suggestion?

**Comment by jameslamb** [comment]
I've created #5387 to track the CI issue. @shiyu1994 let's please continue the conversation there.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [reply python-package/lightgbm/basic.py]: I just applied this suggestion, to hopefully move this PR forward.

**Review by StrikerRUS — APPROVED** [review]

**Pull request merged**
